src/monty/io.py

```diff
@@ -67,105 +121,135 @@ def reverse_readfile(filename: Union[str, Path]) -> Generator[str, str, None]:
-            operation. This is used to determine when to reverse a file
-            in-memory versus seeking portions of a file. For bz2 files,
-            this sets the maximum block size.
+        m_file (File): File stream to read (backwards).
+        blk_size (int): The buffer size in bytes. Defaults to 4096.
+        max_mem (int): The maximum amount of RAM to use in bytes,
+            which determines when to reverse a file in-memory versus
+            seeking segments of a file. For bz2 files, this sets
+            the block size.
 
-    Returns:
-        Generator that yields lines from the file. Behave similarly to the
-        file.readline() function, except the lines are returned from the back
-        of the file.
+    Yields:
+        Lines from the back of the file.
     """
-    # Check if the file stream is a bit stream or not
+    # Generate line ending
+    l_end = _get_line_ending(m_file)
+
+    # Check if the file stream is a buffered text stream
     is_text = isinstance(m_file, io.TextIOWrapper)
 
     try:
         file_size = os.path.getsize(m_file.name)
     except AttributeError:
-        # Bz2 files do not have name attribute. Just set file_size to above
-        # max_mem for now.
+        # Bz2 files do not have "name" attribute.
+        # Just set file_size to max_mem for now.
         file_size = max_mem + 1
 
-    # If the file size is within our desired RAM use, just reverse it in memory
-    # GZip files must use this method because there is no way to negative seek
+    # If the file size is within desired RAM limit, just reverse it in memory.
+    # Gzip files must use this method because there is no way to negative seek.
     # For windows, we also read the whole file.
-    if file_size < max_mem or isinstance(m_file, gzip.GzipFile) or os.name == "nt":
+    if (
+        platform.system() == "Windows"
+        or file_size < max_mem
+        or isinstance(m_file, gzip.GzipFile)
+    ):
         for line in reversed(m_file.readlines()):
-            yield line.rstrip()
+            yield (line if isinstance(line, str) else line.decode("utf-8"))
+
     else:
         if isinstance(m_file, bz2.BZ2File):
-            # for bz2 files, seeks are expensive. It is therefore in our best
-            # interest to maximize the blk_size within limits of desired RAM
-            # use.
+            # For bz2 files, seeking is expensive. It is therefore in our best
+            # interest to maximize the blk_size within RAM usage limit.
             blk_size = min(max_mem, file_size)
 
         buf = ""
         m_file.seek(0, 2)
-        lastchar = m_file.read(1) if is_text else m_file.read(1).decode("utf-8")
+        last_char = m_file.read(1) if is_text else m_file.read(1).decode("utf-8")
 
-        trailing_newline = lastchar == os.linesep
+        trailing_newline = last_char == l_end
 
         while True:
-            newline_pos = buf.rfind(os.linesep)
```

Comment:
Need to rename it because strictly it's not the position of newline, but the end of line